File: crates/brainwires-a2a/src/client/sse.rs
//! SSE (Server-Sent Events) stream parser.

use bytes::Bytes;
use futures::Stream;

use crate::error::A2aError;
use crate::jsonrpc::JsonRpcResponse;
use crate::streaming::StreamEvent;

/// Parse an SSE response body into a stream of `StreamEvent`.
///
/// Expects lines of the form `data: {...}\n\n` where each data line
/// contains a JSON-RPC response with a `StreamEvent` as the result.
pub fn parse_sse_stream(
    body: String,
) -> impl Stream<Item = Result<StreamEvent, A2aError>> {
    async_stream::stream! {
        for line in body.lines() {
            let line = line.trim();
            if let Some(data) = line.strip_prefix("data: ") {
                match serde_json::from_str::<JsonRpcResponse>(data) {
                    Ok(resp) => {
                        if let Some(err) = resp.error {
                            yield Err(err);
                        } else if let Some(result) = resp.result {
                            match serde_json::from_value::<StreamEvent>(result) {
                                Ok(event) => yield Ok(event),
                                Err(e) => yield Err(A2aError::from(e)),
                            }
                        }
                    }
                    Err(e) => {
                        yield Err(A2aError::parse_error(e.to_string()));
                    }
                }
            }
        }
    }
}

/// Parse SSE from a streaming byte source, yielding events as they arrive.
pub fn parse_sse_bytes(
    data: Bytes,
) -> impl Stream<Item = Result<StreamEvent, A2aError>> {
    let text = String::from_utf8_lossy(&data).to_string();
    parse_sse_stream(text)
}

/// Parse an SSE byte stream incrementally (JSON-RPC envelope).
///
/// Reads chunks from a `reqwest::Response::bytes_stream()`, buffers until
/// complete SSE frames (`\n\n` boundaries) are found, then parses each frame.
/// Handles multi-line `data:` fields per the SSE specification.
pub fn parse_sse_byte_stream(
    stream: impl Stream<Item = Result<Bytes, reqwest::Error>> + Send + 'static,
) -> impl Stream<Item = Result<StreamEvent, A2aError>> + Send {
    async_stream::stream! {
        use futures::StreamExt;
        let mut pinned = std::pin::pin!(stream);
        let mut buffer = String::new();

        while let Some(chunk) = pinned.next().await {
            let chunk = match chunk {
                Ok(c) => c,
                Err(e) => {
                    yield Err(A2aError::internal(format!("Stream read error: {e}")));
                    return;
                }
            };
            buffer.push_str(&String::from_utf8_lossy(&chunk));

            // Process complete SSE frames (delimited by \n\n)
            while let Some(boundary) = buffer.find("\n\n") {
                let frame = buffer[..boundary].to_string();
                buffer = buffer[boundary + 2..].to_string();

                if let Some(event) = parse_sse_frame_jsonrpc(&frame) {
                    yield event;
                }
            }
        }

        // Process any remaining data in the buffer
        if !buffer.trim().is_empty() {
            if let Some(event) = parse_sse_frame_jsonrpc(&buffer) {
                yield event;
            }
        }
    }
}

/// Parse an SSE byte stream incrementally (raw REST — no JSON-RPC envelope).
///
/// Like `parse_sse_byte_stream` but expects `data:` lines containing raw
/// `StreamEvent` JSON rather than a JSON-RPC response wrapper.
pub fn parse_sse_rest_byte_stream(
    stream: impl Stream<Item = Result<Bytes, reqwest::Error>> + Send + 'static,
) -> impl Stream<Item = Result<StreamEvent, A2aError>> + Send {
    async_stream::stream! {
        use futures::StreamExt;
        let mut pinned = std::pin::pin!(stream);
        let mut buffer = String::new();

        while let Some(chunk) = pinned.next().await {
            let chunk = match chunk {
                Ok(c) => c,
                Err(e) => {
                    yield Err(A2aError::internal(format!("Stream read error: {e}")));
                    return;
                }
            };
            buffer.push_str(&String::from_utf8_lossy(&chunk));

            while let Some(boundary) = buffer.find("\n\n") {
                let frame = buffer[..boundary].to_string();
                buffer = buffer[boundary + 2..].to_string();

                if let Some(event) = parse_sse_frame_rest(&frame) {
                    yield event;
                }
            }
        }

        if !buffer.trim().is_empty() {
            if let Some(event) = parse_sse_frame_rest(&buffer) {
                yield event;
            }
        }
    }
}

/// Extract the concatenated `data:` payload from an SSE frame.
///
/// Per the SSE spec, multiple `data:` lines are concatenated with newlines.
/// Lines starting with `:` are comments and ignored. `event:`, `id:`, and
/// `retry:` fields are ignored.
fn extract_sse_data(frame: &str) -> Option<String> {
    let mut data_parts: Vec<&str> = Vec::new();

    for line in frame.lines() {
        let line = line.trim_end_matches('\r');
        if let Some(value) = line.strip_prefix("data:") {
            data_parts.push(value.strip_prefix(' ').unwrap_or(value));
        }
        // Ignore event:, id:, retry:, and comments (:)
    }

    if data_parts.is_empty() {
        return None;
    }

    let payload = data_parts.join("\n");
    if payload.is_empty() {
        None
    } else {
        Some(payload)
    }
}

/// Parse an SSE frame with JSON-RPC envelope.
fn parse_sse_frame_jsonrpc(frame: &str) -> Option<Result<StreamEvent, A2aError>> {
    let data = extract_sse_data(frame)?;

    Some(match serde_json::from_str::<JsonRpcResponse>(&data) {
        Ok(resp) => {
            if let Some(err) = resp.error {
                Err(err)
            } else if let Some(result) = resp.result {
                serde_json::from_value::<StreamEvent>(result).map_err(A2aError::from)
            } else {
                return None;
            }
        }
        Err(e) => Err(A2aError::parse_error(e.to_string())),
    })
}

/// Parse an SSE frame with raw StreamEvent JSON (no JSON-RPC envelope).
fn parse_sse_frame_rest(frame: &str) -> Option<Result<StreamEvent, A2aError>> {
    let data = extract_sse_data(frame)?;

    Some(
        serde_json::from_str::<StreamEvent>(&data)
            .map_err(|e| A2aError::parse_error(e.to_string())),
    )
}

File: deno/a2a/sse.ts
/**
 * SSE (Server-Sent Events) stream parser for A2A.
 *
 * Handles both JSON-RPC envelope and raw REST modes.
 * Buffer is capped at 16 MiB to prevent unbounded memory growth.
 * Parses StreamResponse wrapper format per A2A v1.0.
 */

import { A2aError } from "./error.ts";
import type { JsonRpcResponse } from "./jsonrpc.ts";
import type { StreamResponse } from "./streaming.ts";

/** Maximum SSE buffer size (16 MB). */
const MAX_SSE_BUFFER_SIZE = 16 * 1024 * 1024;

/**
 * Extract concatenated `data:` payload from an SSE frame.
 * Per the SSE spec, multiple `data:` lines are concatenated with newlines.
 * Lines starting with `:` are comments and ignored.
 */
function extractSseData(frame: string): string | null {
  const dataParts: string[] = [];
  for (const rawLine of frame.split("\n")) {
    const line = rawLine.replace(/\r$/, "");
    if (line.startsWith("data:")) {
      const value = line.slice(5);
      dataParts.push(value.startsWith(" ") ? value.slice(1) : value);
    }
    // Ignore event:, id:, retry:, and comments (:)
  }
  if (dataParts.length === 0) return null;
  const payload = dataParts.join("\n");
  return payload.length === 0 ? null : payload;
}

/**
 * Parse an SSE frame expecting a JSON-RPC envelope.
 * Extracts the `result` field as a StreamResponse.
 */
function parseSseFrameJsonRpc(frame: string): StreamResponse | A2aError {
  const data = extractSseData(frame);
  if (data === null) {
    return A2aError.parseError("SSE frame contains no data field");
  }

  let resp: JsonRpcResponse;
  try {
    resp = JSON.parse(data) as JsonRpcResponse;
  } catch (e) {
    return A2aError.parseError(String(e));
  }

  if (resp.error) {
    return A2aError.fromJSON(
      resp.error as { code: number; message: string; data?: unknown },
    );
  }

  if (resp.result !== undefined) {
    return resp.result as StreamResponse;
  }

  return A2aError.parseError(
    "JSON-RPC response has neither result nor error",
  );
}

/**
 * Parse an SSE frame expecting raw StreamResponse JSON (REST mode).
 */
function parseSseFrameRest(frame: string): StreamResponse | A2aError {
  const data = extractSseData(frame);
  if (data === null) {
    return A2aError.parseError("SSE frame contains no data field");
  }

  try {
    return JSON.parse(data) as StreamResponse;
  } catch (e) {
    return A2aError.parseError(String(e));
  }
}

/**
 * Parse an SSE byte stream incrementally, yielding StreamResponse values.
 *
 * @param body - ReadableStream from a fetch() response
 * @param mode - "jsonrpc" expects JSON-RPC envelope; "rest" expects raw events
 */
export async function* parseSseStream(
  body: ReadableStream<Uint8Array>,
  mode: "jsonrpc" | "rest" = "jsonrpc",
): AsyncIterable<StreamResponse> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      if (buffer.length > MAX_SSE_BUFFER_SIZE) {
        throw A2aError.internal("SSE stream buffer exceeded maximum size");
      }

      // Process complete SSE frames (delimited by \n\n)
      let pos: number;
      while ((pos = buffer.indexOf("\n\n")) !== -1) {
        const frame = buffer.slice(0, pos);
        buffer = buffer.slice(pos + 2);

        const result = mode === "jsonrpc"
          ? parseSseFrameJsonRpc(frame)
          : parseSseFrameRest(frame);

        if (result instanceof A2aError) {
          throw result;
        }
        yield result;
      }
    }

    // Process any remaining data in the buffer
    if (buffer.trim().length > 0) {
      const result = mode === "jsonrpc"
        ? parseSseFrameJsonRpc(buffer)
        : parseSseFrameRest(buffer);

      if (result instanceof A2aError) {
        throw result;
      }
      yield result;
    }
  } finally {
    reader.releaseLock();
  }
}
